Support verbosity level 0 and 2 in `getblock` RPC command

zzantares opened the pull request:
The verbosity level when calling `getblock` RPC command wasn't configurable,
now it is, but only 0 and 2 are supported which should cover most of the
use-cases. Specifically this change was made to access fees per transaction in
a specified block.

zzantares committed:
Support verbosity level 0 and 2 in `getblock` RPC command

The verbosity level when calling `getblock` RPC command wasn't configurable,
now it is, but only 0 and 2 are supported which should cover most of the
use-cases. Specifically this change was made to access fees per transaction in
a specified block.

GambolingPangolin commented on bitcoind-rpc/src/Bitcoin/Core/RPC/Blockchain.hs:
```diff
@@ -258,7 +264,54 @@ type BlockchainRpc =
+    , getBlockV2MerkleRoot :: Hash256
+    , getBlockV2Time :: UTCTime
+    , getBlockV2Nonce :: Word32
+    , getBlockV2Txs :: [(TxHash, Maybe Word64)]
```
IMO this should include the full transaction e.g. `[(TxHash, Tx, Maybe Word64)]`.  It should be up to the consumer to lose information, not the library author.  After making this change, we could make `getBlockBlock` always succeed and remove the export `getBlock'` (since it would be convenient to compose `getBlock` and `getBlockBlock`).

zzantares replied:
I made the change, but I wasn't sure about what you meant by "we could make `getBlockBlock` always succeed", you mean to change it as `getBlockBlock :: BlockHash -> BitcoindClient Block`?

zzantares replied:
Not sure if it's just me, but GitHub is not showing the new commit? but it is there in the branch https://github.com/bitnomial/bitcoind-rpc/commit/0f10b132a9203c9ffb9baa9965d5587ebe78ab68 weird 🤔

GambolingPangolin replied:
I'm not seeing it either.  Anyway, I mean `getBlockBlock :: GetBlockResponse -> Block` by reassembling the fields in `Block` from the response.

zzantares replied:
Ok I have updated the code in 22da62f8d, it should address this, feel free to unresolve conversation otherwise 👍🏼.

zzantares committed:
Include `Tx` details in `getblock` response

We should still include details about each transaction in the response from
the `getblock` RPC call when called with a verbosity level of 2.

zzantares committed:
Reconstruct to `Block` from `GetBlockResponse`

As pointed out we can extract a `Block` from the `getblock` RPC command
response so that downstream users can rely on `getBlockBlock` to get it from a
response.

zzantares committed:
Update doc comment about the `getBlockBlock` helper

GambolingPangolin reviewed (commented):
Looks good!  Would you run `hlint` once and address any issues in your patch?

zzantares committed:
hlint fixes

zzantares commented:
> Looks good! Would you run `hlint` once and address any issues in your patch?

@GambolingPangolin it should be fixed now.

GambolingPangolin commented on bitcoind-rpc/src/Bitcoin/Core/RPC/Blockchain.hs:
```diff
@@ -32,37 +34,39 @@ module Bitcoin.Core.RPC.Blockchain (
 import Data.Word (Word16, Word32, Word64)
-import Haskoin.Block (Block, BlockHash, BlockHeight, hexToBlockHash)
+import Haskoin.Block (Block (..), BlockHash, BlockHeight, hexToBlockHash)
+import qualified Haskoin.Block as Haskoin
```
Changing this to `import qualified Haskoin as H` would make the import consistent with the rest of the codebase.  I think `Haskoin` exports the `Block` accessors.

zzantares replied:
Ok I checked and yes it does export the `Block` accessors, I've changed it now, good suggestion 👍🏼.

zzantares committed:
Change qualified import to be more consistent with codebase
